transport/http/agent_test.go

```diff
@@ -6,299 +6,256 @@ import (
-				2,
-			)
-
-			s.appliedFunc = func(a *Agent) error {
-				expP1 := s.expectedPeers[0]
-				expP2 := s.expectedPeers[1]
-				expP3 := s.expectedPeers[2]
-				rndP1 := hostport.PeerIdentifier("localhost:9888")
-				rndP2 := hostport.PeerIdentifier("localhost:9883")
-				rndSub1 := peertest.NewMockSubscriber(mockCtrl)
-				rndSub2 := peertest.NewMockSubscriber(mockCtrl)
-				rndSub3 := peertest.NewMockSubscriber(mockCtrl)
-
-				// exp1: Defer a bunch of Releases
-				a.RetainPeer(expP1.identifier, rndSub1)
-				defer a.ReleasePeer(expP1.identifier, rndSub1)
-				a.RetainPeer(expP1.identifier, expP1.subscribers[0])
-				a.RetainPeer(expP1.identifier, rndSub2)
-				defer a.ReleasePeer(expP1.identifier, rndSub2)
-				a.RetainPeer(expP1.identifier, expP1.subscribers[1])
-
-				// exp2: Retain a subscriber, release it, then retain it again
-				a.RetainPeer(expP2.identifier, expP2.subscribers[0])
-				a.RetainPeer(expP2.identifier, expP2.subscribers[1])
-				a.ReleasePeer(expP2.identifier, expP2.subscribers[0])
-				a.ReleasePeer(expP2.identifier, expP2.subscribers[1])
-				a.RetainPeer(expP2.identifier, expP2.subscribers[0])
-				a.RetainPeer(expP2.identifier, expP2.subscribers[1])
-
-				// exp3: Retain release a Peer
-				a.RetainPeer(expP3.identifier, rndSub3)
-				a.ReleasePeer(expP3.identifier, rndSub3)
-				a.RetainPeer(expP3.identifier, expP3.subscribers[0])
-				a.RetainPeer(expP3.identifier, expP3.subscribers[1])
-
-				// rnd1: retain/release on random sub
-				a.RetainPeer(rndP1, rndSub1)
-				a.ReleasePeer(rndP1, rndSub1)
-
-				// rnd2: retain/release on already used subscriber
-				a.RetainPeer(rndP2, expP1.subscribers[0])
-				a.ReleasePeer(rndP2, expP1.subscribers[0])
-
-				return nil
-			}
-			return
-		}(),
+		{
+			msg:         "one retain",
+			identifiers: []string{"i1"},
+			subscriberDefs: []SubscriberDefinition{
+				{ID: "s1"},
+			},
+			actions: []AgentAction{
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s1", ExpectedPeerID: "i1"},
+			},
+			expectedPeers: []peerExpectation{
+				{id: "i1", subscribers: []string{"s1"}},
+			},
+		},
+		{
+			msg:         "one retain one release",
+			identifiers: []string{"i1"},
+			subscriberDefs: []SubscriberDefinition{
+				{ID: "s1"},
+			},
+			actions: []AgentAction{
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s1", ExpectedPeerID: "i1"},
+				ReleaseAction{InputIdentifierID: "i1", InputSubscriberID: "s1"},
+			},
+		},
+		{
+			msg:         "three retains",
+			identifiers: []string{"i1"},
+			subscriberDefs: []SubscriberDefinition{
+				{ID: "s1"},
+				{ID: "s2"},
+				{ID: "s3"},
+			},
+			actions: []AgentAction{
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s1", ExpectedPeerID: "i1"},
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s2", ExpectedPeerID: "i1"},
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s3", ExpectedPeerID: "i1"},
```

Comment:
Not an issue for now but: The attribute names don't need to be this verbose.
They're namespaced to the struct they belong to so they can be generic and
short.